artichoke-backend/src/extn/core/env/env.rb

```diff
@@ -52,63 +52,94 @@ def include?(name)
     to_h.key?(name)
   end
 
+  def invert
+    to_h.invert
+  end
+
   def key(value)
     to_h.key(value)
   end
 
+  def keys
+    to_h.keys
+  end
+
   def length
     to_h.length
   end
 
-  def size
-    to_h.size
+  def rehash
+    nil
   end
 
-  def keys
-    to_h.keys
-  end
+  def shift
+    envs = to_h
 
-  def rehash
-    nil
+    return nil if envs.nil? || envs.empty?
+
+    name, value = envs.shift
+
+    self[name] = nil
+    [name, value]
   end
 
-  def to_a
-    to_h.to_a
+  def size
+    to_h.size
   end
 
-  def to_s
-    'ENV'
+  def replace(hash)
+    clear
+
+    hash.each do |key, value|
+      self[key] = value
+    end
+
+    self
   end
 
-  def value?(name)
-    to_h.value?(name)
+  def select
+    return to_enum(:select) unless block_given?
+
+    to_h.select { |key, value| yield key, value }
   end
 
-  def values
-    to_h.values
+  def select!
+    return to_enum(:select!) unless block_given?
+
+    old_env = to_h
+
+    new_envs = old_env.select { |key, value| yield key, value }
+
+    # returns nil if no changes were made
+    return nil if new_envs == old_env
+
+    clear
```

Comment:
there is a race here where the env is empty for a bit. The docs say that this method should delete the keys that are not present in new env.

> Deletes every environment variable where the block evaluates to false.

Does this work?

```ruby
env = to_h
# collect all the keys where the block evaluates to false
to_remove = env.reject do |key, value|
  yield key, value
end

return nil if to_remove.empty?

to_remove.each do |key, _|
  delete(key)
end

self
```

This method can be aliased to `filter!`.
This method body should be used to implement `keep_if` and `select!` should be implemented with `keep_if`.